packages/vow/src/retrier.js

```diff
@@ -0,0 +1,137 @@
+import { M } from '@endo/patterns';
+import { PromiseWatcherI } from '@agoric/base-zone';
+import { isUpgradeDisconnection } from '@agoric/internal/src/upgrade-api.js';
+import { heapVowE, prepareVowTools } from '../vat.js';
+import { VowShape, toPassableCap } from './vow-utils.js';
+
+/**
+ * @import { Zone } from '@agoric/base-zone'
+ * @import {VowTools} from './tools.js'
+ */
+
+const RetrierI = M.interface('Retrier', {
+  retry: M.call().returns(),
+  getVow: M.call().returns(VowShape),
+  cancel: M.call(M.error()).returns(),
+});
+
+const RetrierShape = M.remotable('retrier');
+
+const RetrierAdminI = M.interface('RetrierAdmin', {
+  // modeled on getFlowForOutcomeVow
+  getRetrierForOutcomeVow: M.call(VowShape).returns(M.opt(RetrierShape)),
+});
+
+/**
+ * @param {Zone} zone
+ * @param {VowTools} [vowTools]
+ */
+export const prepareRetrierTools = (zone, vowTools = prepareVowTools(zone)) => {
+  const { makeVowKit, watch } = vowTools;
+  const retrierForOutcomeVowKey = zone.mapStore('retrierForOutcomeVow', {
+    keyShape: M.remotable('toPassableCap'),
+    valueShape: RetrierShape,
+  });
+
+  const makeRetrierKit = zone.exoClassKit(
+    'Retrier',
+    {
+      retrier: RetrierI,
+      watcher: PromiseWatcherI,
+    },
+    (target, optVerb, args) => {
+      const { vow, resolver } = makeVowKit();
+      return harden({
+        target,
```

Comment:
`target` should support a promise, which is not durably storable. Instead I believe we need a watchHandler for the target, and use `watch(target)` on init to get at the final target. It would also allow us to transparently handle vows as targets too :)